File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder()); 

        int N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(br.readLine());

            if (maxHeap.size() == minHeap.size()) {
                maxHeap.add(num);
            } else {
                minHeap.add(num);
            }
            
            if (!minHeap.isEmpty() && !maxHeap.isEmpty() && maxHeap.peek() > minHeap.peek()) {
                int temp = maxHeap.poll();
                maxHeap.add(minHeap.poll());
                minHeap.add(temp);
            }
            
            bw.write(maxHeap.peek() + "\n");
        }
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.py
import heapq
import sys

min_heap = [] 
max_heap = []  

def add_number(num):
    """
    minheap을 이용해 maxheap 사용.
    maxheap이 minheap과 같거나 하나만 더 길게 설정하여 중앙값 계산
    그렇게 해서 홀 수개이거나 짝수 개든 maxheap출력하면 된다.
    """
    if not max_heap or num <= -max_heap[0]:
        heapq.heappush(max_heap, -num)
    else:
        heapq.heappush(min_heap, num)

    if len(min_heap) > len(max_heap):
        heapq.heappush(max_heap, -heapq.heappop(min_heap))
    elif len(max_heap) > len(min_heap)+ 1:
        heapq.heappush(min_heap, -heapq.heappop(max_heap))

N = int(sys.stdin.readline())
for _ in range(N):
    num = int(sys.stdin.readline())
    add_number(num)
    print(-max_heap[0])
